src/main/java/com/google/solutions/ml/api/vision/VisionAnalyticsPipeline.java

```diff
@@ -51,37 +109,216 @@ public static void main(String[] args) throws Exception {
+            batchedImageURIs.apply(
+                "Annotate Images",
+                ParDo.of(new AnnotateImagesDoFn(options.getFeatures())));
+
+    Map<String, AnnotateImageResponseProcessor> processors = configureProcessors(options);
+
+    PCollection<KV<BQDestination, TableRow>> annotationOutcome =
+        annotatedImages.apply(
+            "Process Annotations",
+            ParDo.of(ProcessImageResponseDoFn.create(ImmutableSet.copyOf(processors.values()))));
+
+    annotationOutcome.apply("Write To BigQuery", new BigQueryDynamicWriteTransform(
+        BQDynamicDestinations.builder()
+            .projectId(options.getVisionApiProjectId())
+            .datasetId(options.getDatasetName())
+            .tableNameToTableDetailsMap(
+                tableNameToTableDetailsMap(processors)).build())
+    );
+
+    collectBatchStatistics(batchedImageURIs, options);
 
     return p.run();
   }
+
+  /**
+   * Collect the statistics on batching the requests. The results are published to a metric. If
+   * {@link VisionAnalyticsPipelineOptions#isCollectBatchData()} is true the batch data is saved to
+   * BigQuery table "batch_info".
+   */
+  static void collectBatchStatistics(PCollection<Iterable<String>> batchedImageURIs,
+      VisionAnalyticsPipelineOptions options) {
+
+    PCollection<TableRow> batchInfo = batchedImageURIs
+        .apply("Collect Batch Stats", ParDo.of(new DoFn<Iterable<String>, TableRow>() {
+          private static final long serialVersionUID = 1L;
+
+          @ProcessElement
+          public void processElement(@Element Iterable<String> element, BoundedWindow window,
+              OutputReceiver<TableRow> out, ProcessContext context) {
+            int size = Iterables.size(element);
+            batchSizeDistribution.update(size);
+            if (context.getPipelineOptions().as(VisionAnalyticsPipelineOptions.class)
+                .isCollectBatchData()) {
+              TableRow row = new TableRow();
+              row.put("window", window.toString());
+              row.put("timestamp", ProcessorUtils.getTimeStamp());
+              row.put("size", size);
+              List<String> items = new ArrayList<>();
+              element.forEach(items::add);
+              row.put("items", items);
+
+              out.output(row);
+            }
+          }
+        }));
+    if (!options.isCollectBatchData()) {
+      return;
+    }
+    batchInfo.apply(
+        BigQueryIO.writeTableRows()
+            .to(new TableReference().setProjectId(options.getVisionApiProjectId())
+                .setDatasetId(options.getDatasetName()).setTableId("batch_info"))
+            .withWriteDisposition(WriteDisposition.WRITE_APPEND)
+            .withoutValidation()
+            .withClustering()
+            .ignoreInsertIds()
+            .withSchema(new TableSchema().setFields(ImmutableList.of(
+                new TableFieldSchema().setName("window").setType("STRING"),
```

Comment:
If possible, I'd recommend using constants for all the values passed to `setType()` instead of raw strings. Although I'm not sure if the bigquery Java SDK already comes with those constants pre-defined.

Reply:
Couldn't find predefined constants in JDK. The JDK is probably auto-generated and the source probably doesn't define it in a way to produce enumeration. Created constants in the repo and updated all references.